pkg/vulnsrc/openeuler/openeuler.go

```diff
@@ -0,0 +1,246 @@
+	}
+)
+
+type VulnSrc struct {
+	dbc db.Operation
+}
+
+func NewVulnSrc() VulnSrc {
+	return VulnSrc{
+		dbc: db.Config{},
+	}
+}
+
+func (vs VulnSrc) Name() types.SourceID {
+	return source.ID
+}
+
+func (vs VulnSrc) Update(dir string) error {
+	log.Println("Saving openEuler CVRF")
+	var cvrfs []Cvrf
+	rootDir := filepath.Join(dir, "vuln-list", eulerDir)
+	err := utils.FileWalk(rootDir, func(r io.Reader, path string) error {
+		var cvrf Cvrf
+		if err := json.NewDecoder(r).Decode(&cvrf); err != nil {
+			return xerrors.Errorf("failed to decode openEuler CVRF JSON: %w %+v", err, cvrf)
+		}
+		cvrfs = append(cvrfs, cvrf)
+		return nil
+	})
+	if err != nil {
+		return xerrors.Errorf("error in openEuler CVRF walk: %w", err)
+	}
+
+	if err = vs.save(cvrfs); err != nil {
+		return xerrors.Errorf("error in openEuler CVRF save: %w", err)
+	}
+
+	return nil
+}
+
+func (vs VulnSrc) save(cvrfs []Cvrf) error {
+	err := vs.dbc.BatchUpdate(func(tx *bolt.Tx) error {
+		return vs.commit(tx, cvrfs)
+	})
+	if err != nil {
+		return xerrors.Errorf("error in batch update: %w", err)
+	}
+	return nil
+}
+
+func (vs VulnSrc) commit(tx *bolt.Tx, cvrfs []Cvrf) error {
+	for _, cvrf := range cvrfs {
+		affectedPkgs := getAffectedPackages(cvrf.ProductTree)
+		if len(affectedPkgs) == 0 {
+			continue
+		}
+
+		for _, affectedPkg := range affectedPkgs {
+			advisory := types.Advisory{
+				FixedVersion: affectedPkg.Package.FixedVersion,
+			}
+
+			if err := vs.dbc.PutDataSource(tx, affectedPkg.OSVer, source); err != nil {
+				return xerrors.Errorf("failed to put data source: %w", err)
+			}
+			if err := vs.dbc.PutAdvisoryDetail(tx, cvrf.Tracking.ID, affectedPkg.Package.Name,
+				[]string{affectedPkg.OSVer}, advisory); err != nil {
+				return xerrors.Errorf("unable to save %s CVRF: %w", affectedPkg.OSVer, err)
+			}
+		}
+
+		var references []string
+		for _, ref := range cvrf.References {
+			references = append(references, ref.URL)
+		}
+
+		severity := types.SeverityUnknown
+		for _, cvuln := range cvrf.Vulnerabilities {
+			for _, threat := range cvuln.Threats {
+				sev := severityFromThreat(threat.Severity)
+				if severity < sev {
+					severity = sev
+				}
+			}
+		}
+
+		vuln := types.VulnerabilityDetail{
+			References:  references,
+			Title:       cvrf.Title,
+			Description: getDetail(cvrf.Notes),
+			Severity:    severity,
+		}
+		if err := vs.dbc.PutVulnerabilityDetail(tx, cvrf.Tracking.ID, source.ID, vuln); err != nil {
+			return xerrors.Errorf("failed to save openEuler CVRF vulnerability: %w", err)
+		}
+
+		// for optimization
+		if err := vs.dbc.PutVulnerabilityID(tx, cvrf.Tracking.ID); err != nil {
+			return xerrors.Errorf("failed to save the vulnerability ID: %w", err)
+		}
+	}
+	return nil
+}
+
+func getAffectedPackages(productTree ProductTree) []AffectedPackage {
+	var pkgs []AffectedPackage
+
+	for _, branch := range productTree.Branches {
+		if branch.Type != "Package Arch" || branch.Name != "aarch64" {
```

Comment:
What about `amd64`? Do we need to check this arch?

Reply:
OK, I declared this in comment. 
It doesn't matter whether the packages are `x86_64` or `aarch64`,  because the number and content of those two kinds of packages are the same. The packages of `aarch64` are chosen because it is more typical for openEuler.

On the other hand, if the packages of `x86_64` are chosen, the result will be the same as those of `aarch64`.

Please see
(aarch64) https://github.com/aquasecurity/vuln-list/blob/24612eef3897977e4671fa4d4488b60da2bc96df/openeuler/2024/openEuler-SA-2024-1001.json#L72

(x86_64)  https://github.com/aquasecurity/vuln-list/blob/24612eef3897977e4671fa4d4488b60da2bc96df/openeuler/2024/openEuler-SA-2024-1001.json#L109

Reply:
I did a little research.
Number of `x86_64` and `aarch64` packages is not equal:
aarch64:
![Screenshot from 2024-07-11 09-27-16](https://github.com/aquasecurity/trivy-db/assets/91113035/9b59c8ed-a935-41bb-84e9-f6e67c739981)
x86_64:
![Screenshot from 2024-07-11 09-43-34](https://github.com/aquasecurity/trivy-db/assets/91113035/4301ed70-04e9-472f-943f-64e69125349b)

> The packages of aarch64 are chosen because it is more typical for openEuler.

We can start from `aarch64`, but perhaps we will need to split `x86_64` and `aarch64`.

___ 

Also i think we need to include `noarch` pkgs (`aarch64` doesn't include `noarch` pkgs (https://github.com/aquasecurity/vuln-list/blob/24612eef3897977e4671fa4d4488b60da2bc96df/openeuler/2023/openEuler-SA-2023-1410.json#L283)) .

IIUC `aarch64` contains `src` package, right?

Reply:
Thanks for your analysis! 
I think we might have to save the arch information of all packages so that we can accurately determine whether there are security vulnerabilities.

`src` pkgs contain the source code but not be installed in openEuler, only pkgs with arch `aarch64`, `x86_64` or `noarch` are installed in openEuler

Reply:
> only pkgs with arch aarch64, x86_64 or noarch are installed in openEuler

Okay, can you add `noarch` packages?

Reply:
Yes, I'm doing this now

Reply:
I just confirmed this, some pkgs may only support `x86_64` or `aarch64`. So we must save the arch information, also include `noarch`

Reply:
We can update this logic and add arches.
You can look at `rocky.go` as example.

Reply:
Thanks, let me see this

Reply:
I have changed the code to save the arches for each package.